### PR TITLE
Make BlobStoreRepository Aware of ClusterState

### PR DESCRIPTION
**Pull request by original-brownbear**
This is a preliminary to #49060.

It does not introduce any substantial behavior change to how the blob store repository
operates. What it does is to add all the infrastructure changes around passing the cluster service
to the blob store, associated test changes and a best effort approach to tracking the latest repository
generation on all nodes from cluster state updates. This brings a slight improvement to the consistency by which non-master nodes (or master directly after a failover) will be able to determine the latest
repository generation. It does not however do any tricky checks for the situation after a repository operation (create, delete or cleanup) that could theoretically be used to get even greater accuracy to keep this change simple.
This change does not in any way alter the behavior of the blobstore repository other than adding a better "guess" for the value of the latest repo generation and is mainly intended to isolate the actual logical change to how the repository operates in #49060

### CODE REVIEW

**Comment by original-brownbear** [comment]
@ywelsch @tlrx I know that the practical improvement to resiliency from this change is pretty small (see PR description ... we're always picking up the best known state before an operation), but I think it should take a lot of the pain out of reviewing #49060 by containing almost all non-functional changes in that PR. Let me know what you think :)

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: let's remove this extra condition (`bestGenerationFromCS == RepositoryData.EMPTY_REPO_GEN`), it's an optimization which does not mattter I think and could hurt us later.

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: would it make sense for these 3 customs to have entries implement an interface that exposes two methods (repository() and getRepositoryStateId()), which will allow simplifying the code here.

**Review by ywelsch — CHANGES_REQUESTED** [review]
Left some comments. Thanks!

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: No actually that's not true I think. We have the sitaution where if you abort a snapshot, the delete entry will be created with `currentRepoGeneration + 1` since that's what the repo will be at when the delete actually runs. That's why I added this.

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: Well, currently there's just one entry here at all times. I don't think that will change before this logic gets replaced by the more inolved logic in #49060 and I'm not sure if and when we move to parallel snapshot taking, that those will in fact have different repository generations set (in my prototype for parallel ops they wouldn't have at least ...).  Could just assert that the count of snapshots in progress is always `1` here? :)

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: Fixed both points in  991f3683bd7 I think :)

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: I thought about that initially but then dropped the idea because that would've meant leaking the specifics of `BlobStoreRepository` into `RepositoriesService`.
But come to think of it now ... doing this would also save a bit of tricky of code in #49060 to get the initialization of the repo on non-master nodes right.
Let's do it -> will do :)

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: ... this has become irrelevant due to the discussion below of not having the CS application happen in the repo class.

File: plugins/repository-azure/src/test/java/org/elasticsearch/repositories/azure/AzureRepositorySettingsTests.java
Comment (tlrx) [inline plugins/repository-azure/src/test/java/org/elasticsearch/repositories/azure/AzureRepositorySettingsTests.java]: Maybe use `BlobStoreTestUtil.mockClusterService()` here too?

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java
Comment (tlrx) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: > I wonder if `RepositoriesService` should update the relevant repository about changes to their snapshots.

I agree, that seems to be the right thing to do. I'm also not super happy to have BlobStoreRepository have its own lifecycle.

**Review by tlrx — COMMENTED** [review]
This looks good already, I'm waiting for Yannick's feedback to be addressed before LGTM :)

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: Could I warm your heart for cfb104f maybe? :)
I think in the case of the repositories it's better to pass the whole `ClusterState` to the repo instead of figuring out what parts go to what repo in `RepositoriesService`. Otherwise we needlessly waste cycles on the repos that don't require the cluster state (e.g. read-only repos, ccr, ...). Also, in #49060 the parts of the CS that need to be inspected will change, but we may want to keep the current logic from here as BwC fallback which is much easier if we just pass the full `ClusterState` down.
Also, I realized that I actually had to move the cluster state application method to `Repository` because of `FilterRepository` (and upcoming encrypted repo wrapper etc.) so this didn't leak any blob store repo specifics to `RepositoriesService` after all :)

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: Done in https://github.com/elastic/elasticsearch/commit/cfb104f9926f477a6a0fa8a47a7bbd83268f708b ... thanks for this suggestion, this actually will allow simplifying a number of spots in a follow-up where we check all three of these `Custom` values for the presence of a certain repo :)

File: plugins/repository-azure/src/test/java/org/elasticsearch/repositories/azure/AzureRepositorySettingsTests.java
Comment (original-brownbear) [reply plugins/repository-azure/src/test/java/org/elasticsearch/repositories/azure/AzureRepositorySettingsTests.java]: Right 6ddfaf0c701 :)

File: server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java]: This isn't strickly necessary with the changes just now, but maybe fine to keep it here since we discovered it here and I think we should def. cleanup on a failed `start` out of principle?

File: server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java
Comment (tlrx) [inline server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java]: I think we need to catch a potential RepositoryException here, log it and continue to update the others repositories.

**Review by tlrx — APPROVED** [review]
LGTM, thanks Armin. I left a small comment where I think we should be defensive just in case an exception is thrown while updating the repositories states.

File: server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java]: I wonder, looking at the code above we aren't defensive around `closeRepository` either (though that might be called in a loop as well) but catch (which has a TODO about it) on `createRepository`.
Maybe we should rather wrap this in a 

```java
try {
    repo.updateState(state);
} catch (Exception e) {
   assert false;
   throw e;
}
```

and do the same for `closeRepository`? Could do it in a follow up and remove that todo while we're at it :)

Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java]: My main contention to just catch and log here would be that once the state update is important for the proper functioning of blob store repositories, then what would an uncaught exception even mean? (imo it would mean that the repo can't be used for writes any longer ... but that would be something the repo would have to set in its internal state when handling exceptions)

**Comment by original-brownbear** [comment]
@tlrx as just discussed fixed the exposing of repositories before update in  https://github.com/elastic/elasticsearch/pull/49639/commits/c257b0e83e9ae6f1a714fd2fff52eef2ec6e7c9a (thanks for catching this!)

**Comment by tlrx** [comment]
> @tlrx as just discussed fixed the exposing of repositories before update in [c257b0e](https://github.com/elastic/elasticsearch/commit/c257b0e83e9ae6f1a714fd2fff52eef2ec6e7c9a) (thanks for catching this!)

Thanks Armin! As discussed I think we should have a look at the other issues in the applyClusterState() method (like maybe only create + update repositories before closing the existing ones) and that could be done in a follow up PR.

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/1 (unrelated transport test failure)

File: server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java]: The bigger issue I think is to get rid of the condition:
```
 if ((oldMetaData == null && newMetaData == null) || (oldMetaData != null && oldMetaData.equals(newMetaData))) {
```

which currently compares the current with the previous cluster state. If anything went wrong applying the previous cluster state, we will not update the repo again.

File: server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java]: why not just put the following into the catch block: `IOUtils.closeWhileHandlingException(repository);`.
No need for success flag and/or finally block

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: I don't understand this condition. Why are we not just taking the max of all the entries that we see? Also, I suppose that we don't allow concurrent operations right now, so we assume that we have only one of these 3 metadata for the current repo?

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: `mapToLong(RepositoryOperation::repositoryStateId).max().orElse(RepositoryData.EMPTY_REPO_GEN);`?

**Review by ywelsch — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: There is the one special case of aborting a snapshot where you can have an in progress snapshot and a delete. The delete will contain the future repository generation for after the snapshot finished -> we can't use that. Thats why I added this condition.

File: server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java]: Right ... fixed now :)

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: Jup that looks nicer :) Fixed now

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/bwc (seems like a messed up version got build ...)
Jenkins run elasticsearch-ci/default-distro

**Comment by original-brownbear** [comment]
Somethings messed up with the BwC tests here (some version constant test), not related to my changes.

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: Can we fix that situation? And can you add a comment to that effect? The condition as it stands here today is very unintuitive.

Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: > Can we fix that situation?

Yes, once we have #49060 we can clean this up when implementing concurrent operations on the repo. That will require the whole business of associating an operation with a strict repo generation to go away anyway.

=> Added a comment for now :)

**Review by ywelsch — APPROVED** [review]
LGTM

**Comment by original-brownbear** [comment]
Thanks Yannick + Tanguy!

**Pull request merged**
